File: examples/python/feeabs_v1beta1_query_messages.py
#!/usr/bin/env python3
"""
xion.feeabs.v1beta1 Query Messages — Protobuf Examples

This file demonstrates how to construct, encode, and decode
query messages from the Fee Abstraction Module (xion.feeabs.v1beta1).

Queries covered:
- QueryHostChainConfig: Get configuration for a specific host chain
- QueryAllHostChainConfig: List all host chain configurations
- QueryOsmosisArithmeticTwap: Get TWAP from Osmosis
- QueryFeeabsModuleBalances: Get module account balances
"""

import sys
import os

# Add the python types to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../python/types'))

from xion.feeabs.v1beta1 import query_pb2 as feeabs_query
from google.protobuf.json_format import MessageToJson


# =============================================================================
# QueryHostChainConfigRequest / Response
# =============================================================================
# Get configuration for a specific host chain.
# Type URL: /xion.feeabs.v1beta1.QueryHostChainConfigRequest

def query_host_chain_config_examples():
    print('=== QueryHostChainConfig Examples ===\n')

    # 1. Construct request
    request = feeabs_query.QueryHostChainConfigRequest()
    request.ibc_denom = 'ibc/27394FB092D2ECCD...'  # IBC denom hash

    print('1. Constructed QueryHostChainConfigRequest:')
    print(f'   ibc_denom: {request.ibc_denom}')
    print(f'   typeUrl: /xion.feeabs.v1beta1.QueryHostChainConfigRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # 3. Decode request
    decoded_request = feeabs_query.QueryHostChainConfigRequest()
    decoded_request.ParseFromString(encoded_request)
    print(f'\n3. Decoded ibc_denom: {decoded_request.ibc_denom}')

    # === Response ===
    print('\n--- Response ---')

    response = feeabs_query.QueryHostChainConfigResponse()
    response.host_chain_config.ibc_denom = 'ibc/27394FB092D2ECCD...'
    response.host_chain_config.osmosis_pool_token_denom_in = 'uatom'
    response.host_chain_config.pool_id = 1
    response.host_chain_config.status = 1  # HostChainFeeAbsStatus

    print('4. Constructed Response:')
    print(f'   host_chain_config.ibc_denom: {response.host_chain_config.ibc_denom}')
    print(f'   host_chain_config.osmosis_pool_token_denom_in: {response.host_chain_config.osmosis_pool_token_denom_in}')
    print(f'   host_chain_config.pool_id: {response.host_chain_config.pool_id}')
    print(f'   host_chain_config.status: {response.host_chain_config.status}')

    # 5. JSON format
    json_str = MessageToJson(response)
    print(f'\n5. JSON format:')
    print(f'   {json_str}')

    return request, response


# =============================================================================
# QueryAllHostChainConfigRequest / Response
# =============================================================================
# List all host chain configurations.
# Type URL: /xion.feeabs.v1beta1.QueryAllHostChainConfigRequest

def query_all_host_chain_config_examples():
    print('\n=== QueryAllHostChainConfig Examples ===\n')

    # 1. Construct request (no parameters needed)
    request = feeabs_query.AllQueryHostChainConfigRequest()

    print('1. Constructed AllQueryHostChainConfigRequest:')
    print('   (no parameters)')
    print(f'   typeUrl: /xion.feeabs.v1beta1.AllQueryHostChainConfigRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = feeabs_query.AllQueryHostChainConfigResponse()
    
    config1 = response.all_host_chain_config.add()
    config1.ibc_denom = 'ibc/27394FB092D2ECCD...'
    config1.osmosis_pool_token_denom_in = 'uatom'
    config1.pool_id = 1

    config2 = response.all_host_chain_config.add()
    config2.ibc_denom = 'ibc/ABC123DEF456...'
    config2.osmosis_pool_token_denom_in = 'uusdc'
    config2.pool_id = 2

    print('3. Constructed Response:')
    print(f'   all_host_chain_config count: {len(response.all_host_chain_config)}')
    for i, config in enumerate(response.all_host_chain_config):
        print(f'   [{i}] ibc_denom: {config.ibc_denom[:30]}...')
        print(f'       pool_id: {config.pool_id}')

    return request, response


# =============================================================================
# QueryOsmosisArithmeticTwapRequest / Response
# =============================================================================
# Get TWAP (Time-Weighted Average Price) from Osmosis.
# Type URL: /xion.feeabs.v1beta1.QueryOsmosisArithmeticTwapRequest

def query_osmosis_arithmetic_twap_examples():
    print('\n=== QueryOsmosisArithmeticTwap Examples ===\n')

    # 1. Construct request
    request = feeabs_query.QueryOsmosisArithmeticTwapRequest()
    request.ibc_denom = 'ibc/27394FB092D2ECCD...'

    print('1. Constructed QueryOsmosisArithmeticTwapRequest:')
    print(f'   ibc_denom: {request.ibc_denom}')
    print(f'   typeUrl: /xion.feeabs.v1beta1.QueryOsmosisArithmeticTwapRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = feeabs_query.QueryOsmosisArithmeticTwapResponse()
    response.arithmetic_twap = '1.234567890000000000'  # Price as decimal string

    print('3. Constructed Response:')
    print(f'   arithmetic_twap: {response.arithmetic_twap}')
    print(f'   (This is the time-weighted average price)')

    return request, response


# =============================================================================
# QueryFeeabsModuleBalancesRequest / Response
# =============================================================================
# Get the fee abstraction module account balances.
# Type URL: /xion.feeabs.v1beta1.QueryFeeabsModuleBalancesRequest

def query_feeabs_module_balances_examples():
    print('\n=== QueryFeeabsModuleBalances Examples ===\n')

    # 1. Construct request (no parameters needed)
    request = feeabs_query.QueryFeeabsModuleBalacesRequest()

    print('1. Constructed QueryFeeabsModuleBalacesRequest:')
    print('   (no parameters)')
    print(f'   typeUrl: /xion.feeabs.v1beta1.QueryFeeabsModuleBalacesRequest')

    # 2. Encode request
    encoded_request = request.SerializeToString()
    print(f'\n2. Encoded request: {len(encoded_request)} bytes')

    # === Response ===
    print('\n--- Response ---')

    response = feeabs_query.QueryFeeabsModuleBalacesResponse()
    
    bal1 = response.balances.add()
    bal1.denom = 'uxion'
    bal1.amount = '1000000000'

    bal2 = response.balances.add()
    bal2.denom = 'ibc/27394FB092D2ECCD...'
    bal2.amount = '500000000'

    response.address = 'xion1feeabsmoduleaddress...'

    print('3. Constructed Response:')
    print(f'   address: {response.address}')
    print(f'   balances count: {len(response.balances)}')
    for bal in response.balances:
        print(f'     - {bal.amount} {bal.denom[:30]}...')

    return request, response


# =============================================================================
# Run All Examples
# =============================================================================

def main():
    print('╔════════════════════════════════════════════════════════════╗')
    print('║  xion.feeabs.v1beta1 Query Messages — Protobuf Examples    ║')
    print('╚════════════════════════════════════════════════════════════╝\n')

    query_host_chain_config_examples()
    query_all_host_chain_config_examples()
    query_osmosis_arithmetic_twap_examples()
    query_feeabs_module_balances_examples()

    print('\n═══════════════════════════════════════════════════════════════')
    print('Type URL Reference (Requests):')
    print('  QueryHostChainConfigRequest:        /xion.feeabs.v1beta1.QueryHostChainConfigRequest')
    print('  AllQueryHostChainConfigRequest:     /xion.feeabs.v1beta1.AllQueryHostChainConfigRequest')
    print('  QueryOsmosisArithmeticTwapRequest:  /xion.feeabs.v1beta1.QueryOsmosisArithmeticTwapRequest')
    print('  QueryFeeabsModuleBalacesRequest:    /xion.feeabs.v1beta1.QueryFeeabsModuleBalacesRequest')
    print('\nREST Endpoints:')
    print('  GET /xion/feeabs/v1beta1/host_chain_config/{ibc_denom}')
    print('  GET /xion/feeabs/v1beta1/all_host_chain_config')
    print('  GET /xion/feeabs/v1beta1/osmosis_arithmetic_twap/{ibc_denom}')
    print('  GET /xion/feeabs/v1beta1/feeabs_module_balances')
    print('═══════════════════════════════════════════════════════════════')


if __name__ == '__main__':
    main()

File: examples/rust/examples/feeabs_v1beta1_query_messages.rs
//! xion.feeabs.v1beta1 Query Messages — Protobuf Examples
//!
//! This file demonstrates how to construct, encode, and decode
//! query messages from the Fee Abstraction Module (xion.feeabs.v1beta1).
//!
//! Queries covered:
//! - QueryOsmosisArithmeticTwap: Get TWAP price for an IBC token
//! - QueryFeeabsModuleBalances: Get module account balances
//! - QueryHostChainConfig: Get configuration for a specific IBC token
//! - QueryAllHostChainConfig: List all host chain configurations
//!
//! Fee Abstraction allows users to pay transaction fees in non-native tokens
//! by swapping them to native tokens via Osmosis cross-chain swaps.

use prost::Message;

use xion_types::types::xion_feeabs_v1beta1::{
    HostChainFeeAbsConfig, HostChainFeeAbsStatus, QueryAllHostChainConfigRequest,
    QueryAllHostChainConfigResponse, QueryFeeabsModuleBalancesRequest,
    QueryFeeabsModuleBalancesResponse, QueryHostChainConfigRequest, QueryHostChainConfigResponse,
    QueryOsmosisArithmeticTwapRequest, QueryOsmosisArithmeticTwapResponse,
};

use xion_types::types::cosmos_base_v1beta1::Coin;

// =============================================================================
// QueryOsmosisArithmeticTwap
// =============================================================================
// Get the Time-Weighted Average Price (TWAP) for an IBC token.
// Endpoint: GET /fee-abstraction/feeabs/v1/osmosis-arithmetic-twap/{ibc_denom}
// Type URL: /xion.feeabs.v1beta1.QueryOsmosisArithmeticTwapRequest

fn query_osmosis_arithmetic_twap_examples() {
    println!("=== QueryOsmosisArithmeticTwap Examples ===\n");

    // 1. Construct request
    let request = QueryOsmosisArithmeticTwapRequest {
        ibc_denom: "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2"
            .to_string(), // ATOM
    };

    println!("1. Constructed QueryOsmosisArithmeticTwapRequest:");
    println!("   ibc_denom: {}", request.ibc_denom);

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request =
        QueryOsmosisArithmeticTwapRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded ibc_denom: {}", decoded_request.ibc_denom);

    // === Response ===
    println!("\n--- Response ---");

    // TWAP is returned as a decimal string (LegacyDec)
    // Example: "8.500000000000000000" means 1 ATOM = 8.5 XION
    let response = QueryOsmosisArithmeticTwapResponse {
        arithmetic_twap: "8.500000000000000000".to_string(),
    };

    println!("4. Constructed Response:");
    println!("   arithmetic_twap: {}", response.arithmetic_twap);
    println!(
        "   Interpretation: 1 IBC token = {} native tokens",
        response.arithmetic_twap.parse::<f64>().unwrap_or(0.0)
    );

    // 5. Encode/decode response
    let encoded_response = response.encode_to_vec();
    let decoded_response =
        QueryOsmosisArithmeticTwapResponse::decode(encoded_response.as_slice()).unwrap();
    println!(
        "\n5. Decoded arithmetic_twap: {}",
        decoded_response.arithmetic_twap
    );
}

// =============================================================================
// QueryFeeabsModuleBalances
// =============================================================================
// Get the balances held by the fee abstraction module account.
// Endpoint: GET /fee-abstraction/feeabs/v1/module-balances
// Type URL: /xion.feeabs.v1beta1.QueryFeeabsModuleBalancesRequest

fn query_feeabs_module_balances_examples() {
    println!("\n=== QueryFeeabsModuleBalances Examples ===\n");

    // 1. Construct request (no parameters needed)
    let request = QueryFeeabsModuleBalancesRequest {};

    println!("1. Constructed QueryFeeabsModuleBalancesRequest:");
    println!("   (no parameters)");

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // === Response ===
    println!("\n--- Response ---");

    let response = QueryFeeabsModuleBalancesResponse {
        balances: vec![
            Coin {
                denom: "uxion".to_string(),
                amount: "1000000000000".to_string(), // 1M XION
            },
            Coin {
                denom: "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2"
                    .to_string(),
                amount: "50000000".to_string(), // 50 ATOM
            },
        ],
        address: "xion1feeabsmoduleaddress...".to_string(),
    };

    println!("3. Constructed Response:");
    println!("   address: {}", response.address);
    println!("   balances:");
    for coin in &response.balances {
        println!("     - {} {}", coin.amount, &coin.denom[..20.min(coin.denom.len())]);
    }

    // 4. Encode/decode response
    let encoded_response = response.encode_to_vec();
    let decoded_response =
        QueryFeeabsModuleBalancesResponse::decode(encoded_response.as_slice()).unwrap();
    println!(
        "\n4. Decoded balances count: {}",
        decoded_response.balances.len()
    );
    println!("   Decoded address: {}", decoded_response.address);
}

// =============================================================================
// QueryHostChainConfig
// =============================================================================
// Get the configuration for a specific IBC token.
// Endpoint: GET /fee-abstraction/feeabs/v1/host-chain-config/{ibc_denom}
// Type URL: /xion.feeabs.v1beta1.QueryHostChainConfigRequest

fn query_host_chain_config_examples() {
    println!("\n=== QueryHostChainConfig Examples ===\n");

    // 1. Construct request
    let request = QueryHostChainConfigRequest {
        ibc_denom: "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2"
            .to_string(),
    };

    println!("1. Constructed QueryHostChainConfigRequest:");
    println!("   ibc_denom: {}", request.ibc_denom);

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // 3. Decode request
    let decoded_request =
        QueryHostChainConfigRequest::decode(encoded_request.as_slice()).unwrap();
    println!("\n3. Decoded ibc_denom: {}", decoded_request.ibc_denom);

    // === Response ===
    println!("\n--- Response ---");

    let response = QueryHostChainConfigResponse {
        host_chain_config: Some(HostChainFeeAbsConfig {
            ibc_denom: "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2"
                .to_string(),
            osmosis_pool_token_denom_in: "uatom".to_string(),
            pool_id: 1,
            status: HostChainFeeAbsStatus::Updated as i32,
        }),
    };

    println!("4. Constructed Response:");
    if let Some(ref config) = response.host_chain_config {
        println!("   host_chain_config.ibc_denom: {}...", &config.ibc_denom[..40]);
        println!(
            "   host_chain_config.osmosis_pool_token_denom_in: {}",
            config.osmosis_pool_token_denom_in
        );
        println!("   host_chain_config.pool_id: {}", config.pool_id);
        println!("   host_chain_config.status: {} (UPDATED)", config.status);
    }

    // 5. Encode/decode response
    let encoded_response = response.encode_to_vec();
    let decoded_response =
        QueryHostChainConfigResponse::decode(encoded_response.as_slice()).unwrap();
    if let Some(config) = decoded_response.host_chain_config {
        println!("\n5. Decoded pool_id: {}", config.pool_id);
    }
}

// =============================================================================
// QueryAllHostChainConfig
// =============================================================================
// Get all host chain configurations.
// Endpoint: GET /fee-abstraction/feeabs/v1/all-host-chain-config
// Type URL: /xion.feeabs.v1beta1.QueryAllHostChainConfigRequest

fn query_all_host_chain_config_examples() {
    println!("\n=== QueryAllHostChainConfig Examples ===\n");

    // 1. Construct request (no parameters needed)
    let request = QueryAllHostChainConfigRequest {};

    println!("1. Constructed QueryAllHostChainConfigRequest:");
    println!("   (no parameters)");

    // 2. Encode request
    let encoded_request = request.encode_to_vec();
    println!("\n2. Encoded request: {} bytes", encoded_request.len());

    // === Response ===
    println!("\n--- Response ---");

    let response = QueryAllHostChainConfigResponse {
        all_host_chain_config: vec![
            HostChainFeeAbsConfig {
                ibc_denom:
                    "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2"
                        .to_string(),
                osmosis_pool_token_denom_in: "uatom".to_string(),
                pool_id: 1,
                status: HostChainFeeAbsStatus::Updated as i32,
            },
            HostChainFeeAbsConfig {
                ibc_denom:
                    "ibc/D189335C6E4A68B513C10AB227BF1C1D38C746766278BA3EEB4FB14124F1D858"
                        .to_string(),
                osmosis_pool_token_denom_in: "uusdc".to_string(),
                pool_id: 678,
                status: HostChainFeeAbsStatus::Updated as i32,
            },
            HostChainFeeAbsConfig {
                ibc_denom:
                    "ibc/498A0751C798A0D9A389AA3691123DADA57DAA4FE165D5C75894505B876BA6E4"
                        .to_string(),
                osmosis_pool_token_denom_in: "uosmo".to_string(),
                pool_id: 1,
                status: HostChainFeeAbsStatus::Outdated as i32,
            },
        ],
    };

    println!("3. Constructed Response:");
    println!(
        "   all_host_chain_config count: {}",
        response.all_host_chain_config.len()
    );
    for (i, config) in response.all_host_chain_config.iter().enumerate() {
        let status_name = match config.status {
            1 => "UPDATED",
            2 => "OUTDATED",
            3 => "FROZEN",
            _ => "UNSPECIFIED",
        };
        println!("\n   [{}] {}:", i, config.osmosis_pool_token_denom_in);
        println!("       ibc_denom: {}...", &config.ibc_denom[..30]);
        println!("       pool_id: {}", config.pool_id);
        println!("       status: {}", status_name);
    }

    // 4. Encode/decode response
    let encoded_response = response.encode_to_vec();
    let decoded_response =
        QueryAllHostChainConfigResponse::decode(encoded_response.as_slice()).unwrap();
    println!(
        "\n4. Decoded config count: {}",
        decoded_response.all_host_chain_config.len()
    );
}

// =============================================================================
// Understanding Fee Abstraction Flow
// =============================================================================

fn fee_abstraction_flow_examples() {
    println!("\n=== Understanding Fee Abstraction Flow ===\n");

    println!("Fee Abstraction allows users to pay fees in non-native tokens.\n");

    println!("Flow:");
    println!("  1. User wants to pay fees in ATOM (IBC token)");
    println!("  2. Query TWAP to get current ATOM/XION price");
    println!("  3. FeeAbs module swaps ATOM → XION via Osmosis");
    println!("  4. Swapped XION is used to pay the actual fee\n");

    println!("Requirements:");
    println!("  • Host chain config must exist for the IBC token");
    println!("  • Config status must be UPDATED (TWAP is current)");
    println!("  • FeeAbs module must have sufficient balance");
    println!("  • Osmosis pool must have liquidity\n");

    println!("IBC Denom Format:");
    println!("  ibc/HASH");
    println!("  where HASH = SHA256(path/channel-id/denom)\n");

    println!("Common IBC Tokens:");
    println!("  • ATOM: ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2");
    println!("  • USDC: ibc/D189335C6E4A68B513C10AB227BF1C1D38C746766278BA3EEB4FB14124F1D858");
    println!("  • OSMO: ibc/498A0751C798A0D9A389AA3691123DADA57DAA4FE165D5C75894505B876BA6E4");
}

// =============================================================================
// Run All Examples
// =============================================================================

fn main() {
    println!("╔════════════════════════════════════════════════════════════╗");
    println!("║  xion.feeabs.v1beta1 Query Messages — Protobuf Examples    ║");
    println!("╚════════════════════════════════════════════════════════════╝\n");

    query_osmosis_arithmetic_twap_examples();
    query_feeabs_module_balances_examples();
    query_host_chain_config_examples();
    query_all_host_chain_config_examples();
    fee_abstraction_flow_examples();

    println!("\n═══════════════════════════════════════════════════════════════");
    println!("Type URL Reference (Requests):");
    println!("  QueryOsmosisArithmeticTwapRequest:  /xion.feeabs.v1beta1.QueryOsmosisArithmeticTwapRequest");
    println!("  QueryFeeabsModuleBalancesRequest:   /xion.feeabs.v1beta1.QueryFeeabsModuleBalancesRequest");
    println!("  QueryHostChainConfigRequest:        /xion.feeabs.v1beta1.QueryHostChainConfigRequest");
    println!("  QueryAllHostChainConfigRequest:     /xion.feeabs.v1beta1.QueryAllHostChainConfigRequest");
    println!("\nREST Endpoints:");
    println!("  GET /fee-abstraction/feeabs/v1/osmosis-arithmetic-twap/{{ibc_denom}}");
    println!("  GET /fee-abstraction/feeabs/v1/module-balances");
    println!("  GET /fee-abstraction/feeabs/v1/host-chain-config/{{ibc_denom}}");
    println!("  GET /fee-abstraction/feeabs/v1/all-host-chain-config");
    println!("═══════════════════════════════════════════════════════════════");
}
